### PR TITLE
Authority Migration Cache Writing

### PR DESCRIPTION
**Pull request by RPangrle**
* Write new tokens only to the preferred cache location
* Tokens should be tombstoned in the correct location regardless of preferred cache location

### CODE REVIEW

File: ADAL/src/cache/ADTokenCacheAccessor.m [hunk 1]
```diff
@@ -342,52 +359,52 @@ - (void)removeItemFromCache:(ADTokenCacheItem *)cacheItem
...
-            [_dataSource addOrUpdateItem:existing correlationId:correlationId error:nil];
-            removed = YES;
-        }
+        return;
     }
     
-    if (!removed)
+    NSUUID* correlationId = [context correlationId];
+    
+    ADTokenCacheItem* existing = [_dataSource getItemWithKey:cacheKey
```
Comment (jasoncoolmax) [inline ADAL/src/cache/ADTokenCacheAccessor.m]: This call will not hit the "fallback read" mechanism, is it possible that it cannot find the item we want to find?

Comment (RPangrle) [reply ADAL/src/cache/ADTokenCacheAccessor.m]: This is the code that tombstones the existing item. We want it tombstoning the existing item we used, not writing out to the preferred location :)

File: ADAL/src/cache/ADTokenCacheItem.m [hunk 2]
```diff
@@ -30,13 +30,17 @@
 #import "ADTokenCacheItem+Internal.h"
 
 @implementation ADTokenCacheItem
+{
+    NSString *_storageAuthority;
```
Comment (partnerinflight) [inline ADAL/src/cache/ADTokenCacheItem.m]: MINOR Rename this field "_storageAuthority" to match the regular expression: ^[a-z][a-zA-Z0-9]*$ [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS116)

File: ADAL/src/cache/ADTokenCacheItem.m [hunk 3]
```diff
@@ -90,6 +94,14 @@ - (id)copyWithZone:(NSZone*) zone
...
 - (ADTokenCacheKey*)extractKey:(ADAuthenticationError* __autoreleasing *)error
 {
+    if (_storageAuthority)
+    {
+        return [ADTokenCacheKey keyWithAuthority:_storageAuthority
+                                        resource:_resource
+                                        clientId:_clientId
+                                           error:error];
+    }
+
```
Comment (partnerinflight) [inline ADAL/src/cache/ADTokenCacheItem.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/validation/ADAuthorityValidation.m [hunk 4]
```diff
@@ -326,6 +326,25 @@ - (NSURL *)networkUrlForAuthority:(NSURL *)authority
...
 }
 
+- (NSURL *)cacheUrlForAuthority:(NSURL *)authority
+                        context:(id<ADRequestContext>)context
+{
+    if ([ADHelpers isADFSInstanceURL:authority])
+    {
+        return authority;
+    }
+
```
Comment (partnerinflight) [inline ADAL/src/validation/ADAuthorityValidation.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/validation/ADAuthorityValidation.m [hunk 5]
```diff
@@ -326,6 +326,25 @@ - (NSURL *)networkUrlForAuthority:(NSURL *)authority
...
+        return authority;
+    }
+    
+    NSURL *url = [_aadCache cacheUrlForAuthority:authority];
+    if (!url)
+    {
+        AD_LOG_WARN(@"No cached preferred_cache for authority", context.correlationId, nil);
+        return authority;
+    }
+
```
Comment (partnerinflight) [inline ADAL/src/validation/ADAuthorityValidation.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/validation/ADAuthorityValidation.m [hunk 6]
```diff
@@ -326,6 +326,25 @@ - (NSURL *)networkUrlForAuthority:(NSURL *)authority
...
+    }
+    
+    NSURL *url = [_aadCache cacheUrlForAuthority:authority];
+    if (!url)
+    {
+        AD_LOG_WARN(@"No cached preferred_cache for authority", context.correlationId, nil);
+        return authority;
+    }
+    
+
```
Comment (partnerinflight) [inline ADAL/src/validation/ADAuthorityValidation.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/cache/ADTokenCacheAccessor.m [hunk 7]
```diff
@@ -342,52 +359,52 @@ - (void)removeItemFromCache:(ADTokenCacheItem *)cacheItem
...
-    ADTelemetryCacheEvent* event = [[ADTelemetryCacheEvent alloc] initWithName:AD_TELEMETRY_EVENT_TOKEN_CACHE_DELETE
-                                                                       context:context];
     
-    [event setSpeInfo:cacheItem.speInfo];
-    [[ADTelemetry sharedInstance] stopEvent:[context telemetryRequestId] event:event];
+    if (!existing || ![refreshToken isEqualToString:existing.refreshToken])
+    {
+        return;
+    }
+
```
Comment (partnerinflight) [inline ADAL/src/cache/ADTokenCacheAccessor.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/cache/ADTokenCacheAccessor.m [hunk 8]
```diff
@@ -342,52 +359,52 @@ - (void)removeItemFromCache:(ADTokenCacheItem *)cacheItem
         return;
     }
     
-    NSUUID* correlationId = [context correlationId];
+
```
Comment (partnerinflight) [inline ADAL/src/cache/ADTokenCacheAccessor.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/cache/ADTokenCacheAccessor.m [hunk 9]
```diff
@@ -87,6 +87,9 @@ - (ADTokenCacheItem *)getItemForUser:(NSString *)userId
                                                       userId:userId
                                                correlationId:[context correlationId]
                                                        error:&adError];
+        item.storageAuthority = item.authority;
+        item.authority = _authority;
+
```
Comment (partnerinflight) [inline ADAL/src/cache/ADTokenCacheAccessor.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/cache/ADTokenCacheAccessor.m [hunk 10]
```diff
@@ -332,6 +335,20 @@ - (void)updateCacheToItem:(ADTokenCacheItem *)cacheItem
...
+                context:(id<ADRequestContext>)context
+                  error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
+{
+    NSURL *oldAuthority = [NSURL URLWithString:item.authority];
+    NSURL *newAuthority = [[ADAuthorityValidation sharedInstance] cacheUrlForAuthority:oldAuthority context:context];
+    
+    item.authority = [newAuthority absoluteString];
+    BOOL ret = [_dataSource addOrUpdateItem:item correlationId:context.correlationId error:error];
+    item.authority = [oldAuthority absoluteString];
+
```
Comment (partnerinflight) [inline ADAL/src/cache/ADTokenCacheAccessor.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/cache/ADTokenCacheAccessor.m [hunk 11]
```diff
@@ -342,52 +359,52 @@ - (void)removeItemFromCache:(ADTokenCacheItem *)cacheItem
...
-                [_dataSource addOrUpdateItem:broadItem correlationId:correlationId error:nil];
-            }
-        }
+        existing = [_dataSource getItemWithKey:[cacheKey mrrtKey]
+                                        userId:cacheItem.userInformation.userId
+                                 correlationId:correlationId
+                                         error:nil];
     }
-    ADTelemetryCacheEvent* event = [[ADTelemetryCacheEvent alloc] initWithName:AD_TELEMETRY_EVENT_TOKEN_CACHE_DELETE
-                                                                       context:context];
```
Comment (partnerinflight) [inline ADAL/src/cache/ADTokenCacheAccessor.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/cache/ADTokenCacheAccessor.m [hunk 12]
```diff
@@ -342,52 +359,52 @@ - (void)removeItemFromCache:(ADTokenCacheItem *)cacheItem
...
-            AD_LOG_VERBOSE_F(@"Token cache store", correlationId, @"Tombstoning cache for resource: %@", cacheItem.resource);
-            //update tombstone property before update the tombstone in cache
-            [existing makeTombstone:@{ @"correlationId" : [correlationId UUIDString],
-                                       @"errorDetails" : [error errorDetails],
-                                       @"protocolCode" : [error protocolCode] }];
-            [_dataSource addOrUpdateItem:existing correlationId:correlationId error:nil];
-            removed = YES;
-        }
+        return;
     }
```
Comment (partnerinflight) [inline ADAL/src/cache/ADTokenCacheAccessor.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

**Comment by partnerinflight** [comment]
SonarQube analysis reported 11 issues
* ![MINOR][MINOR] 11 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: ADAL/src/cache/ADTokenCacheAccessor.m (hunk 1 above)
Comment (jasoncoolmax) [reply ADAL/src/cache/ADTokenCacheAccessor.m]: Forget what I said, I didn't notice `extractKey` is also changed when I made that comment :)

**Comment by partnerinflight** [comment]
SonarQube analysis reported 10 issues
* ![MINOR][MINOR] 10 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: ADAL/src/cache/ADTokenCacheAccessor.m [hunk 13]
```diff
@@ -332,6 +335,20 @@ - (void)updateCacheToItem:(ADTokenCacheItem *)cacheItem
...
+- (BOOL)addOrUpdateItem:(nonnull ADTokenCacheItem *)item
+                context:(id<ADRequestContext>)context
+                  error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
+{
+    NSURL *oldAuthority = [NSURL URLWithString:item.authority];
+    NSURL *newAuthority = [[ADAuthorityValidation sharedInstance] cacheUrlForAuthority:oldAuthority context:context];
+    
+    item.authority = [newAuthority absoluteString];
+    BOOL ret = [_dataSource addOrUpdateItem:item correlationId:context.correlationId error:error];
+    item.authority = [oldAuthority absoluteString];
```
Comment (oldalton) [inline ADAL/src/cache/ADTokenCacheAccessor.m]: nit: please add a comment here about, because it's difficult to understand without knowing context

Comment (RPangrle) [reply ADAL/src/cache/ADTokenCacheAccessor.m]: 👍

File: ADAL/src/cache/ADTokenCacheAccessor.m [hunk 14]
```diff
@@ -332,6 +335,23 @@ - (void)updateCacheToItem:(ADTokenCacheItem *)cacheItem
...
+    NSURL *oldAuthority = [NSURL URLWithString:item.authority];
+    NSURL *newAuthority = [[ADAuthorityValidation sharedInstance] cacheUrlForAuthority:oldAuthority context:context];
+    
+    // The authority used to retrieve the item over the network can differ from the preferred authority used to
+    // cache the item. As it would be awkward to cache an item using an authority other then the one we store
+    // it with we switch it out before saving it to cache.
+    item.authority = [newAuthority absoluteString];
+    BOOL ret = [_dataSource addOrUpdateItem:item correlationId:context.correlationId error:error];
+    item.authority = [oldAuthority absoluteString];
+
```
Comment (partnerinflight) [inline ADAL/src/cache/ADTokenCacheAccessor.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

**Pull request merged**